### PR TITLE
[Kernel] Skip reserved writes in MLA KV cache

### PR DESCRIPTION
**Pull request by tanth47**
## Motivation

Related to #36207.

SGLang reserves physical KV-cache slot 0 as a finite padding source for CUDA
graph and DP-attention padding. Generic KV stores protect that invariant after
#32477, but MLA-specific writers bypass `store_cache()` and still accept
`loc == 0` as a write destination.

Undefined graph/DP-padding rows can therefore overwrite slot 0 with NaNs. A
later padded attention row may read the poisoned slot and propagate NaNs into
attention output and model logits. During the GLM-5.2 investigation, affected
requests produced short nonsensical output followed by one repeated token.

## Root-cause boundary

This PR fixes the SGLang-owned MLA writers:

| Writer | Covered here |
| --- | ---: |
| BF16 Triton MLA writer | Yes |
| FP8 quantized MLA writer | Yes |
| MLA scale writer | Yes |
| CUDA TMA/JIT MLA writer | Yes |
| AITER fused GLM rope/cat/cache writer | No |

The fused ROCm path through `aiter.fused_qk_rope_cat_and_cache_mla` writes KV
internally and then invokes attention with `save_kv_cache=False`. It bypasses
the writers changed here and requires a separate API-safe integration fix.
This PR should therefore not be interpreted as resolving the complete GLM ROCm
path by itself.

## Modifications

- Add `reserved_skip_index=0` to the Triton MLA data, FP8-quantized, and
  scale-buffer writers.
- Apply the same runtime guard to the CUDA TMA bulk-store writer.
- Preserve static launch shapes and DCP ownership behavior; no Python-side
  compaction is introduced.
- Support `reserved_skip_index=-1` to opt out, matching `store_cache()`
  semantics.
- Add focused CUDA and AMD regression tests.

## Deterministic reproduction

Initialize slot 0 to a finite sentinel and invoke an MLA writer with
`loc=[0, 2]`. The source row mapped to slot 0 may be NaN because it represents
padding, while the row mapped to slot 2 is valid.

Without this patch:

- slot 0 is overwritten by the padding row;
- the valid positive slot is written normally.

With this patch:

- slot 0 remains byte-for-byte unchanged;
- the valid positive slot is still written;
- passing `reserved_skip_index=-1` restores the legacy behavior explicitly.

The tests exercise this contract directly instead of depending on
uninitialized allocator contents.

## Accuracy tests

- Added kernel tests for BF16 MLA writes with int32 and int64 locations.
- Added FP8-quantized and scale-buffer reserved-slot tests.
- Added CUDA TMA/JIT writer coverage.
- Added the explicit opt-out path.
- Python syntax compilation and `git diff --check` pass locally.
- Local GPU tests were not run for this PR; the focused tests are registered
  for CUDA and AMD CI.

The broader real-weight GLM investigation separately demonstrated the same
invariant violation in the fused AITER writer: `slot_mapping=[0]` with a NaN
padding row changed a finite slot 0 into 576 NaNs. That result motivates the
system-level issue but is outside this PR's implementation scope.

## Speed tests and profiling

Not run. The kernels add one scalar comparison to the existing store predicate
without changing launch grids or tensor shapes.

## Checklist

- [x] Format code according to the project style.
- [x] Add focused CUDA and AMD tests.
- [x] Document the API behavior in the writer docstrings.
- [x] Preserve existing behavior through an explicit opt-out.
- [x] Clarify the fused AITER scope boundary.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #32921116235](https://github.com/sgl-project/sglang/actions/runs/32921116235)
Latest PR Test (Extra): :x: [Run #32921116036](https://github.com/sgl-project/sglang/actions/runs/32921116036)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #32921116122](https://github.com/sgl-project/sglang/actions/runs/32921116122)

### CODE REVIEW

**Comment by tanth47** [comment]
Scope clarification for the ROCm path: this PR protects the MLA KV-cache writers owned by SGLang. The fused ROCm path through `aiter.fused_qk_rope_cat_and_cache_mla` writes KV internally, then invokes attention with `save_kv_cache=False`, so it bypasses the writers changed here. That fused AITER writer therefore needs a corresponding reserved-slot fix, which we are tracking separately. This does not change the purpose of this PR: it establishes and enforces the reserved padding-slot invariant across the generic SGLang MLA writer paths.

File: test/registered/kernels/ops/kvcache/test_mla_buffer.py
Comment (kpham-sgl) [inline test/registered/kernels/ops/kvcache/test_mla_buffer.py:1]: This should be merged into `test_set_mla_kv_buffer.py`

**Review by kpham-sgl — COMMENTED** [review]
I think this is mostly an invariant guard? Writing to cache_loc=0 is *fine* if the value is finite.
The actual root cause for NaN might be somewhere else in  ROCm?

**Comment by kpham-sgl** [comment]
@tanth47 please fix lint

File: test/registered/kernels/ops/kvcache/test_mla_buffer.py
Comment (tanth47) [reply test/registered/kernels/ops/kvcache/test_mla_buffer.py:1]: Done — I merged the Triton/FP8/scale cases into `test_set_mla_kv_buffer.py` and removed the standalone test file in 64f51d9a9. I also kept AMD coverage while skipping the CUDA-only TMA cases on ROCm.

**Comment by tanth47** [comment]
Thanks — agreed that this is an invariant guard, not a claim that every finite write to physical slot 0 is invalid.

Within SGLang, slot 0 is the shared reserved padding source. The failure mode in #36207 is that a graph/DP padding row may be non-finite; allowing that row to target slot 0 poisons the reserved source, which a later padded reader can touch. The explicit `reserved_skip_index=-1` opt-out remains for callers that intentionally want slot 0 to be writable.

The fused AITER ROCm writer bypasses the SGLang writers in this PR, so I agree that this PR alone does not fix every ROCm NaN source; that scope boundary is now called out in the PR body and tracked separately in #36207.

I also addressed the lint/review feedback in 40aa45593: the tests are consolidated into `test_set_mla_kv_buffer.py`, the standalone file is removed, and the changed files pass the repo's pinned isort, Black, Ruff, clang-format, codespell, registry lint, bare-pytest-main lint, py_compile, and `git diff --check` locally. CI is rerunning on the new commit.

**Comment by tanth47** [comment]
@kpham-sgl Yes, this is an invariant guard. I don’t think the underlying issue is specific to ROCm, padding rows may be undefined or non-finite regardless of backend, following the #32477. Only the generic cache writer already protects reserved slot 0. This PR simply extends the same protection to the MLA writers.

**Review by kpham-sgl — APPROVED** [review]

**Comment by ormandj** [comment]
I reproduced the reserved-slot overwrite on NVIDIA CUDA as well, not only ROCm.

Setup: DeepSeek-V4-Flash-0731, TP2 on two RTX PRO 6000 Blackwell GPUs (SM120), FP8 KV cache, running my downstream SGLang-based `0.8.1-rc4` build. This configuration reaches `set_mla_kv_buffer_triton`.

I drove that wrapper with a synthetic two-row input: one row containing non-finite values targeting `loc == 0`, and one row containing finite values targeting `loc == 2`.

```text
baseline, without this PR:
slot0_changed=True
slot2_valid_write=True
slot0_nan_counts=fp8:0,scale:4,rope:64

with this PR at 64f51d9a95efb84dba1291b62250360db5eb300f:
slot0_changed=False
slot2_valid_write=True
slot0_nan_counts=fp8:0,scale:0,rope:0
```

Without the PR, the write to reserved slot 0 lands on CUDA and leaves NaNs in the scale and RoPE components. With the PR, slot 0 is untouched. The valid write to `loc == 2` succeeds in both cases, so the guard does not suppress the legitimate write on this path.

This confirms that the unguarded reserved-slot write affects the CUDA FP8 MLA path, not only ROCm. I did not exercise the fused AITER ROCm path, so I make no claim about it. I did not reproduce the PR's reported short nonsensical output followed by one repeated token, so this probe establishes the invariant violation but not a causal link to that user-visible behavior.

The focused upstream GPU tests pass: 8 passed. I also integrated the change into my serving build and ran five repetitions at every decode concurrency against the same build without the PR. Forward-rate changes at C1/C2/C4/C8/C16/C32 ranged from -1.08% to +0.73%, in both directions across concurrencies. Five cache-cold prefill requests at each of 8K/32K/64K/128K ranged from -0.58% to -0.01%. Zero request errors.

**Comment by kpham-sgl** [comment]
Failures unrelated / fixed on main
